doc/sbsetroot.1.in
'\" t
.\"     Title: sbsetroot
.\"    Author: zlice https://github.com/zlice/shynebox
.\" Generator: DocBook XSL Stylesheets vsnapshot <http://docbook.sf.net/>
.\"      Date: 11 October 2023
.\"    Manual: Shynebox Manual
.\"    Source: sbsetroot.txt
.\"  Language: English
.\"
.TH "SBSETROOT" "1" "11 October 2023" "sbsetroot\&.txt" "Shynebox Manual"
.\" -----------------------------------------------------------------
.\" * Define some portability stuff
.\" -----------------------------------------------------------------
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" http://bugs.debian.org/507673
.\" http://lists.gnu.org/archive/html/groff/2009-02/msg00013.html
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"
sbsetroot \- a simple background utility used by the shynebox(1) window manager, originally written for and by the original blackbox(1) window manager\&.
.SH "SYNOPSIS"
.sp
\fBsbsetroot\fR [\-display \fIdisplay\fR] \-solid \fIcolor\fR
.sp
\fBsbsetroot\fR [\-display \fIdisplay\fR] \-mod \fIx\fR \fIy\fR \-fg \fIcolor\fR \-bg \fIcolor\fR
.sp
\fBsbsetroot\fR [\-display \fIdisplay\fR] \-gradient \fItexture\fR \-from \fIcolor\fR \-to \fIcolor\fR
.sp
\fBsbsetroot\fR \-help
.SH "DESCRIPTION"
.sp
\fBsbsetroot(1)\fR is a utility that can control the appearance of the root window in three ways: Either give it a solid color, or write a two color modula pattern to it, or render a gradient texture based on two different colors\&.
.sp
sbsetroot resembles \fBxsetroot(1)\fR in this functionality but it supports multiple\-screen displays, and gradient textures the same way as blackbox or fluxbox does\&. It doesn\(cqt handle cursors\&. sbsetroot was originally part of the Blackbox package and was carried over with the code that became Fluxbox and then Shynebox\&. It was called bsetroot/fbsetroot back in those days\&.
.sp
If any errors are encountered, sbsetroot will use either \fBgxmessage(1)\fR or \fBxmessage(1)\fR to inform the user of errors\&.
.SH "OPTIONS"
.sp
sbsetroot operates in three ways, you must choose one of the first 3 options:
.PP
\fB\-solid\fR \fIcolor\fR
.RS 4
Sets the root window to specified color\&.
.RE
.PP
\fB\-mod\fR \fIx\fR \fIy\fR
.RS 4
Creates a modula pattern\&. You must specify
\fB\-bg\fR
and
\fB\-fg\fR
colors\&.
.RE
.PP
\fB\-gradient\fR \fItexturestring\fR
.RS 4
Renders the specified texture string to the root window\&.
.PP
\fItexturestring\fR may be one of:
.RS 4
\fBHorizontal / Vertical / Diagonal / Crossdiagonal / Pipecross / Elliptic / Rectangle / Pyramid\fR
.sp
Select one of these texture types, they only apply when
\fB\-gradient\fR
is specified\&. You must also specify both a
\fB\-from\fR
and a
\fB\-to\fR
color\&.
.RE
.RE
.PP
\fB\-display\fR \fIdisplay\fR
.RS 4
Tells sbsetroot to connect to the specified display\&.
.RE
.PP
\fB\-bg, \-background\fR \fIcolor\fR
.RS 4
Background color\&. Needed for
\fB\-mod\fR
patterns\&.
.RE
.PP
\fB\-fg, \-foreground\fR \fIcolor\fR
.RS 4
Foreground color\&. Needed for
\fB\-mod\fR
patterns\&.
.RE
.PP
\fB\-from\fR \fIcolor\fR
.RS 4
Start color for rendering textures\&. Needed for
\fB\-gradient\fR
patterns\&.
.RE
.PP
\fB\-to\fR \fIcolor\fR
.RS 4
Ending color for rendering textures\&. Needed for
\fB\-gradient\fR
patterns\&.
.RE
.PP
\fB\-help\fR
.RS 4
Prints version info and short help text\&.
.RE
.SH "AUTHORS"
.sp
This manpage was modified by Curt "Asenchi" Micol <asenchi at asenchi\&.com> for the Fluxbox window manager\&.
.sp
Further updates for fluxbox\-1\&.1\&.2 and conversion to asciidoc format by Jim Ramsay <i\&.am at jimramsay\&.com>
.sp
Updated for shynebox by zlice
.SH "SEE ALSO"
.sp
shynebox(1) sbsetbg(1) xsetroot(1) xmessage(1) gxmessage(1)
.SH "AUTHOR"
.PP
\fBzlice https://github\&.com/zlice/shynebox\fR
.RS 4
Author.
.RE


util/sbsetroot.cc
// Copyright (c) 2023 zlice https://github.com/zlice/shynebox
// Copyright (c) 2002 - 2006 Henrik Kinnunen (fluxgen at fluxbox dot org)
// Copyright (c) 1997 - 2000 Brad Hughes <bhughes at trolltech.com>
//
// Permission is hereby granted, free of charge, to any person obtaining a
// copy of this software and associated documentation files (the "Software"),
// to deal in the Software without restriction, including without limitation
// the rights to use, copy, modify, merge, publish, distribute, sublicense,
// and/or sell copies of the Software, and to permit persons to whom the
// Software is furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL
// THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER

#include "sbsetroot.hh"

#include "../src/tk/I18n.hh"
#include "../src/tk/ImageControl.hh"
#include "../src/tk/Texture.hh"
#include "../src/tk/GContext.hh"
#include "../src/SbRootWindow.hh"

#include <X11/Xatom.h>

#include <cstring>
#include <cstdlib>
#include <cstdio>
#include <iostream>

using std::cout;
using std::cerr;
using std::endl;
using std::string;

sbsetroot::sbsetroot(int argc, char **argv, char *dpy_name)
    : tk::App(dpy_name), m_app_name(argv[0]) {

    pixmap = (Pixmap *) 0;
    screen = DefaultScreen(tk::App::instance()->display());
    grad = fore = back = (char *) 0;

    bool mod = false, sol = false, grd = false;
    int mod_x = 0, mod_y = 0, i = 1;

    img_ctrl = new tk::ImageControl(screen);

    for (; i < argc; i++) {
        string arg = argv[i];
        if (arg == "-help" || arg == "--help" || arg == "-h") {
            usage();

        } else if (arg == "-fg" || arg == "-foreground" ||
                   arg == "--foreground" || arg == "-from" || arg == "--from") {
            if ((++i) >= argc)
                usage(1);
            fore = argv[i];

        } else if (arg == "-bg" || arg == "-background" ||
                   arg == "--background" || arg == "-to" || arg == "--to") {
            if ((++i) >= argc)
                usage(1);
            back = argv[i];

        } else if (arg == "-solid" || arg == "--solid") {
            if ((++i) >= argc)
                usage(1);
            fore = argv[i];
            sol = true;

        } else if (arg == "-mod" || arg == "--mod") {
            if ((++i) >= argc)
                usage();
            mod_x = atoi(argv[i]);
            if ((++i) >= argc)
                usage();
            mod_y = atoi(argv[i]);
            if (mod_x < 1)
                mod_x = 1;
            if (mod_y < 1)
                mod_y = 1;
            mod = true;

        } else if (arg == "-gradient" || arg == "--gradient") {
            if ((++i) >= argc)
                usage();

            grad = argv[i];
            grd = true;

        } else if (arg == "-display" || arg == "--display") {
            // -display passed through tests earlier... we just skip it now
            i++;

        } else
            usage();
    }

    if ((mod + sol + grd) != true) {
        _SB_USES_NLS;
        cerr << _SB_CONSOLETEXT(sbsetroot, MustSpecify,
                    "Error: must specify one of: -solid, -mod, -gradient\n",
                    "user didn't give one of the required options") << endl;

        exit(2);
    }

    if (sol && fore)
        solid();
    else if (mod && mod_x && mod_y && fore && back)
        modula(mod_x, mod_y);
    else if (grd && grad && fore && back)
        gradient();
    else
        usage();

}


sbsetroot::~sbsetroot() {
    XKillClient(display(), AllTemporary);

    if (pixmap) { // should always be true
        XSetCloseDownMode(display(), RetainTemporary);

        delete pixmap;
    }

    delete img_ctrl;
}

/**
   set root pixmap atoms so that apps like
   Eterm and xchat will be able to use
   transparent background
*/
void sbsetroot::setRootAtoms(Pixmap pixmap, int screen) {
    Atom atom_root, atom_eroot, type;
    unsigned char *data_root, *data_eroot;
    int format;
    unsigned long length, after;

    atom_root = XInternAtom(display(), "_XROOTMAP_ID", true);
    atom_eroot = XInternAtom(display(), "ESETROOT_PMAP_ID", true);
    SbRootWindow root(screen);

    // doing this to clean up after old background
    if (atom_root != None && atom_eroot != None) {
        root.property(atom_root, 0L, 1L, false, AnyPropertyType,
                      &type, &format, &length, &after, &data_root);

        if (type == XA_PIXMAP) {
            root.property(atom_eroot, 0L, 1L, False, AnyPropertyType,
                          &type, &format, &length, &after, &data_eroot);

            if (data_root && data_eroot && type == XA_PIXMAP &&
                *((Pixmap *) data_root) == *((Pixmap *) data_eroot)) {

                XKillClient(display(), *((Pixmap *) data_root));
            }
        }
    }

    atom_root = XInternAtom(display(), "_XROOTPMAP_ID", false);
    atom_eroot = XInternAtom(display(), "ESETROOT_PMAP_ID", false);

    if (atom_root == None || atom_eroot == None) {
        _SB_USES_NLS;
        cerr<<_SB_CONSOLETEXT(sbsetroot, NoPixmapAtoms, "Couldn't create pixmap atoms, giving up!", "Couldn't create atoms to point at root pixmap")<<endl;
        exit(1);
    }

    // setting new background atoms
    root.changeProperty(atom_root, XA_PIXMAP, 32, PropModeReplace, (unsigned char *) &pixmap, 1);
    root.changeProperty(atom_eroot, XA_PIXMAP, 32, PropModeReplace, (unsigned char *) &pixmap, 1);

}

/**
   Draws pixmaps with a single color
*/
void sbsetroot::solid() {
    tk::Color c(fore, screen);
    if (! c.isAllocated())
        c.setPixel(BlackPixel(display(), screen));

    SbRootWindow root(screen);

    tk::GContext gc(root);
    gc.setForeground(c);

    pixmap = new Pixmap(XCreatePixmap(display(),
                                      root.window(),
                                      root.width(), root.height(),
                                      root.depth()));

    XFillRectangle(display(), *pixmap, gc.gc(), 0, 0,
                   root.width(), root.height());

    setRootAtoms(*pixmap, screen);

    root.setBackgroundPixmap(*pixmap);
    root.clear();
}

/**
 Draws pixmaps with an 16x16 pattern with
 fg and bg colors.
*/
void sbsetroot::modula(int x, int y) {

    const int s = 16;

    char data[32];
    long pattern = 0;

    int i;

    SbRootWindow root(screen);

    for (i = 0; i < s; i++) {
        pattern <<= 1;
        if ((i % x) == 0)
            pattern |= 0x0001;
    }

    for (i = 0; i < s; i++) {
        if ((i %  y) == 0) {
            data[(i * 2)] = (char) 0xff;
            data[(i * 2) + 1] = (char) 0xff;
        } else {
            data[(i * 2)] = pattern & 0xff;
            data[(i * 2) + 1] = (pattern >> 8) & 0xff;
        }
    }


    Pixmap bitmap, r_bitmap;


    bitmap = XCreateBitmapFromData(display(),
                                   root.window(), data, s, s);

    // bitmap used as tile, needs to have the same depth as background pixmap
    r_bitmap = XCreatePixmap(display(),
                             root.window(), s, s,
                             root.depth());

    tk::Color f(fore, screen), b(back, screen);

    if (! f.isAllocated())
        f.setPixel(WhitePixel(display(), screen));
    if (! b.isAllocated())
        b.setPixel(BlackPixel(display(), screen));

    tk::GContext gc(root);

    gc.setForeground(f);
    gc.setBackground(b);

    // copying bitmap to the one going to be used as tile
    XCopyPlane(display(), bitmap, r_bitmap, gc.gc(),
               0, 0, s, s, 0, 0, 1l);

    gc.setTile(r_bitmap);
    gc.setFillStyle(FillTiled);

    pixmap = new Pixmap(XCreatePixmap(display(),
                                      root.window(),
                                      root.width(), root.height(),
                                      root.depth()));

    XFillRectangle(display(), *pixmap, gc.gc(), 0, 0,
                   root.width(), root.height());

    setRootAtoms(*pixmap, screen);
    root.setBackgroundPixmap(*pixmap);
    root.clear();

    XFreePixmap(display(), bitmap);
    XFreePixmap(display(), r_bitmap);
}

/**
 draws pixmaps with a shynebox texure
*/
void sbsetroot::gradient() {
    // using temporaray pixmap and then copying it to background pixmap, as it'll
    // get crashed somewhere on the way causing apps like XChat chrashing
    // as the pixmap has been destroyed
    Pixmap tmp;
    // we must insert gradient text
    string texture_value = grad ? grad : "solid";
    texture_value.insert(0, "gradient ");
    tk::Texture texture;
    texture.setFromString(texture_value.c_str());

    SbRootWindow root(screen);


    tk::GContext gc(root);
    texture.color().setFromString(fore, screen);
    texture.colorTo().setFromString(back, screen);


    if (! texture.color().isAllocated())
        texture.color().setPixel(WhitePixel(display(), screen));

    if (! texture.colorTo().isAllocated())
        texture.colorTo().setPixel(BlackPixel(display(), screen));

    tmp = img_ctrl->renderImage(root.width(), root.height(), texture);

    pixmap = new Pixmap(XCreatePixmap(display(),
                                      root.window(),
                                      root.width(), root.height(),
                                      root.depth()));


    XCopyArea(display(), tmp, *pixmap, gc.gc(), 0, 0,
              root.width(), root.height(),
              0, 0);

    setRootAtoms(*pixmap, screen);

    root.setBackgroundPixmap(*pixmap);
    root.clear();

    if (! (root.visual()->c_class & 1)) {
        img_ctrl->removeImage(tmp);
        img_ctrl->cleanCache();
    }

}

/**
 Shows information about usage
*/
void sbsetroot::usage(int exit_code) {
    _SB_USES_NLS;
    cout << m_app_name << " 2.3 : (c) 2023 Shynebox - zlice" << endl;
    cout << m_app_name << " 2.3 : (c) 2003-2015 Fluxbox Development Team" << endl;
    cout << m_app_name << " 2.1 : (c) 2002 Claes Nasten" << endl;
    cout << m_app_name << " 2.0 : (c) 1997-2000 Brad Hughes\n" << endl;
    cout << _SB_CONSOLETEXT(sbsetroot, Usage,
                  "  -display <string>        display connection\n"
                  "  -mod <x> <y>             modula pattern\n"
                  "  -foreground, -fg <color> modula foreground color\n"
                  "  -background, -bg <color> modula background color\n\n"
                  "  -gradient <texture>      gradient texture\n"
                  "  -from <color>            gradient start color\n"
                  "  -to <color>              gradient end color\n\n"
                  "  -solid <color>           solid color\n\n"
                  "  -help                    print this help text and exit\n",
                  "sbsetroot usage options") << endl;
    exit(exit_code);
}


int main(int argc, char **argv) {
    char *display_name = (char *) 0;
    int i = 1;

    tk::I18n::init(0);

    for (; i < argc; i++) {
        if (!strcmp(argv[i], "-display") || !strcmp(argv[i], "--display")) {
            // check for -display option

            if ((++i) >= argc) {
                _SB_USES_NLS;
                cerr<<_SB_CONSOLETEXT(main, DISPLAYRequiresArg,
                              "error: '-display' requires an argument",
                              "option requires an argument")<<endl;

                ::exit(1);
            }

            display_name = argv[i];
        }
    }

    try {
        sbsetroot app(argc, argv, display_name);
    } catch (string & error_str) {
        _SB_USES_NLS;
        cerr<<_SB_CONSOLETEXT(Common, Error, "Error", "Error message header")<<": "<<error_str<<endl;
    }

    return (0);
}
